app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+require 'aws-sdk'
+require 'amazon_ssa_support'
+require 'yaml'
+require 'open3'
+
+class AmazonAgentManager
+  include Comparable
+  include Vmdb::Logging
+  attr_accessor :ems, :deploying
+
+  MIQ_SSA   = "MIQ_SSA".freeze
+  RUBY_VERS = "2.3.3".freeze
+  AGENT_AMI = "centos-7.2-hvm*".freeze
+
+  def initialize(ems)
+    @ems = ems
+    @guid = MiqServer.my_server.guid
+
+    # List of active agent ids
+    @alive_agent_ids = []
+
+    # List of all agent ids, include those in power off state.
+    @agent_ids = []
+    @deploying = false
+  end
+
+  def ec2
+    @ec2 ||= ems.connect(:service => 'EC2')
+  end
+
+  def sqs
+    @sqs ||= ems.connect(:service => 'SQS')
+  end
+
+  def s3
+    @s3 ||= ems.connect(:service => 'S3')
+  end
+
+  def iam
+    @iam ||= ems.connect(:service => 'IAM')
+  end
+
+  def <=>(other)
+    @ems.guid <=> other.ems.guid
+  end
+
+  def agent_manager_settings
+    @agent_manager_settings ||= Settings.ems.ems_amazon.try(:agent_manager)
+  end
+
+  def region
+    @ems.provider_region
+  end
+
+  def log_level
+    ll = agent_manager_settings.try(:log_level) || AmazonSsaSupport::DEFAULT_LOG_LEVEL
+    ll.upcase
+  end
+
+  def heartbeat_prefix
+    agent_manager_settings.try(:heartbeat_prefix) || AmazonSsaSupport::DEFAULT_HEARTBEAT_PREFIX
+  end
+
+  def heartbeat_interval
+    agent_manager_settings.try(:heartbeat_interval) || AmazonSsaSupport::DEFAULT_HEARTBEAT_INTERVAL
+  end
+
+  def ssa_bucket
+    @ssa_bucket ||= (agent_manager_settings.try(:bucket_prefix) || AmazonSsaSupport::DEFAULT_BUCKET_PREFIX) + '-' + @ems.guid
+  end
+
+  def request_queue
+    @request_queue ||= (agent_manager_settings.try(:request_queue_prefix) || AmazonSsaSupport::DEFAULT_REQUEST_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_queue
+    @reply_queue ||= (agent_manager_settings.try(:reply_queue_prefix) || AmazonSsaSupport::DEFAULT_REPLY_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_prefix
+    agent_manager_settings.try(:reply_prefix) || AmazonSsaSupport::DEFAULT_REPLY_PREFIX
+  end
+
+  def log_prefix
+    agent_manager_settings.try(:log_prefix) || AmazonSsaSupport::DEFAULT_LOG_PREFIX
+  end
+
+  def agent_ami
+    agent_manager_settings.try(:agent_ami) || AGENT_AMI
+  end
+
+  def ruby_version
+    agent_manager_settings.try(:ruby_version) || RUBY_VERS
+  end
+
+  def userdata_script
+    agent_manager_settings.try(:userdata_script)
```

Comment:
The script is stored in the settings? or is the script's path?  If the latter, it might be clearer if this was named `userdata_script_file` or `userdata_script_path`